File: lua/hob/logs.lua
-- --- hob.logs ---
-- Diagnostics for flows.
--
-- Lines go through the engine so they stay ordered with the rest of the flow
-- and can appear in a trace. `info` and up are visible by default; `debug` and
-- `trace` need the engine's higher verbosity levels.

local hob = require("hob")

local logs = {}

local function emit(level, message)
  return hob.effect("logs", "write", { level = level, msg = message })
end

--- Fine-grained tracing.
function logs.trace(message)
  return emit("trace", message)
end

--- Developer-facing detail.
function logs.debug(message)
  return emit("debug", message)
end

--- Progress worth reading.
function logs.info(message)
  return emit("info", message)
end

--- Something recoverable went wrong.
function logs.warn(message)
  return emit("warn", message)
end

--- Something failed.
function logs.error(message)
  return emit("error", message)
end

return logs

File: src/effect/ops/logs.rs
// --- effect::ops::logs ---
// Arguments of `logs.write`.

use serde::Deserialize;

/// Severity of one log line.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub(crate) enum Level {
    /// Fine-grained tracing, shown under `-vvv`.
    Trace,
    /// Developer-facing detail, shown under `-vv`.
    Debug,
    /// Progress worth reading.
    Info,
    /// Something recoverable went wrong.
    Warn,
    /// Something failed.
    Error,
}

impl Level {
    /// The word a log line is prefixed with.
    pub(crate) fn label(self) -> &'static str {
        match self {
            Self::Trace => "trace",
            Self::Debug => "debug",
            Self::Info => "info",
            Self::Warn => "warn",
            Self::Error => "error",
        }
    }

    /// Severity as a number, compared against the verbosity flag.
    pub(crate) fn rank(self) -> u8 {
        match self {
            Self::Trace => 3,
            Self::Debug => 2,
            Self::Info => 1,
            Self::Warn | Self::Error => 0,
        }
    }
}

/// One log entry.
#[derive(Debug, Deserialize)]
pub(crate) struct Write {
    /// Severity.
    pub(crate) level: Level,
    /// Text of the line.
    pub(crate) msg: String,
}
